src/DataStreams/CsvStream.php

```diff
@@ -0,0 +1,347 @@
+
+    private $writable = false;
+
+    private $headers = null;
+
+    private $hasHeaders = true;
+
+    private $needToWriteHeaders = true;
+
+    private $remnantBuffer = "";
+
+    /**
+     * The character used to enclose string values that might contain the delimiter.
+     * @var string
+     */
+    public $enclosure = "\"";
+
+    /**
+     * The delimiter character, normally a comma or semicolon.
+     *
+     * @var string
+     */
+    public $delimiter = ",";
+
+    /**
+     * The escape character that precedes legitimate occurrences of the enclosure character within the
+     * cell content.
+     *
+     * Note an RFC4180 compliant CSV file will escape the enclosure character by repeating it so for example
+     * the value Summer "Time" becomes Summer ""Time"". In this case you should leave $escapeCharacter null
+     * in which case the RFC behaviour will be adopted.
+     *
+     * @var string|bool
+     */
+    public $escapeCharacter = null;
+
+    public function __construct($filePath)
+    {
+        $this->filePath = $filePath;
+
+        parent::__construct();
+    }
+
+    public function setHeaders($headers)
+    {
+        $this->headers = $headers;
+        $this->hasHeaders = true;
+    }
+
+    private function readHeaders()
+    {
+        $this->close();
+
+        if (!file_exists($this->filePath)) {
+            $this->needToWriteHeaders = true;
+            $this->headers = [];
+
+            return;
+        }
+
+        $this->fileStream = fopen($this->filePath, "r");
+
+        $rawCsvData = $this->readCsvLine();
+
+        $this->headers = $rawCsvData;
+        $this->needToWriteHeaders = false;
+    }
+
+    private function readCsvLine()
+    {
+        if (feof($this->fileStream) && ($this->remnantBuffer == "")) {
+            throw new EndOfStreamException();
+        }
+
+        $inEnclosure = false;
+        $readFullLine = false;
+
+        $values = [];
+
+        $addValue = function (&$value) use (&$values) {
+            $values[] = utf8_encode($value);
+            $value = "";
+        };
+
+        $valueBuffer = "";
+
+        $escapeCharacter = ($this->escapeCharacter !== null) ? $this->escapeCharacter : $this->enclosure;
+
+        while (!$readFullLine) {
+            if (feof($this->fileStream) && ($this->remnantBuffer == "")) {
+                break;
+            }
+
+            // Read the next set of bytes and prepend with any remaining buffer from the last read.
+
+            if (strlen($this->remnantBuffer) < 1024) {
+                $csvData = $this->remnantBuffer . fread($this->fileStream, 1024);
+            } else {
+                $csvData = $this->remnantBuffer;
+            }
+
+            // Be sure to clear the remnant buffer.
+            $this->remnantBuffer = "";
+
+            $csvDataLength = strlen($csvData);
+
+            for ($i = 0; $i < $csvDataLength; $i++) {
+                $byte = $csvData[$i];
+
+                if ($i < $csvDataLength - 1) {
```

Comment:
Yes - original I had an else case on the inner 'if' (hence the 2 if statements) although even now I think it's maybe okay - it makes it easier to understand the context "this bit only applies if we have at least one character left to go" - sometimes when if's are combined it can make them hard to decipher quickly when you come back to the code later.

Good call on the "" empty field. I actually had a test for $inEnclosure but I removed it to try and allow it to capture cases where people totally ignore the file spec (think propertynews feeds) and have unenclosed string values but still use the escape sequence e.g.

some guy,some \"other\" guy,2,

However we can still test for the 'empty' cell scenario separately.